File: app/src/main/java/net/xaethos/trackernotifier/subscribers/ErrorToastSubscriber.java
package net.xaethos.trackernotifier.subscribers;

import android.content.Context;
import android.widget.Toast;

import rx.Subscriber;
import rx.subscriptions.Subscriptions;

public abstract class ErrorToastSubscriber<T> extends Subscriber<T> {

    Context mContext;

    public ErrorToastSubscriber(Context context) {
        mContext = context;
        add(Subscriptions.create(() -> mContext = null));
    }

    @Override
    public void onError(Throwable error) {
        Context context = mContext;
        if (context != null) Toast.makeText(mContext, error.getMessage(), Toast.LENGTH_LONG).show();
    }

    @Override
    public void onCompleted() {
        //noop
    }
}

File: app/src/main/java/net/xaethos/trackernotifier/subscribers/ErrorToastSubscriber.kt
package net.xaethos.trackernotifier.subscribers

import android.content.Context
import android.widget.Toast

import rx.Subscriber
import rx.subscriptions.Subscriptions

abstract class ErrorToastSubscriber<T>(context: Context) : Subscriber<T>() {

    internal var context: Context? = context

    init {
        add(Subscriptions.create { this.context = null })
    }

    override fun onError(error: Throwable) {
        context?.toastError(error)
    }

    override fun onCompleted() {
        //noop
    }
}

fun Context.toastError(error: Throwable) =
        Toast.makeText(this, error.message, Toast.LENGTH_LONG).show()
